.github/workflows/_build.yml

```diff
@@ -0,0 +1,272 @@
+
+  skip-build:
+    runs-on: ubuntu-latest
+    needs: [check-build-exists]
+    if: needs.check-build-exists.outputs.exists == 'true'
+
+    steps:
+      - name: Build Exists, Skip To Test
+        run: |
+          echo A build already exists for version ${{inputs.version_number}}, skipping build job
+
+  build-packages:
+    runs-on: ubuntu-latest
+    needs: [check-build-exists, build-changelog-path]
+    if: needs.check-build-exists.outputs.exists == 'false'
+
+    outputs:
+      finished: ${{ steps.set_success.outputs.finished }}
+
+    steps:
+
+      - name: Check out the repository
+        uses: actions/checkout@v3
+        with:
+          # persist-credentials: false  # TODO: with this false I get an error checking it out, it's what existed in previous release version though
+          ref: ${{ inputs.sha }}
+
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+
+      - name: Install python dependencies
+        run: |
+          pip install --user --upgrade pip
+          pip install --upgrade setuptools wheel twine check-wheel-contents
+          pip --version
+
+      - name: Build distributions
+        run: ./${{ inputs.build_script_path }}
+
+      - name: Show distributions
+        run: ls -lh dist/
+
+      - name: Check distribution descriptions
+        run: |
+          twine check dist/*
+
+      - name: Check wheel contents
+        run: |
+          check-wheel-contents dist/*.whl --ignore W007,W008
+
+      # We want the artifacts available for release and Pypi steps
+      # so we don't have to reach back out to S3
+      - name: upload artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ inputs.version_number }}
+          path: |
+            dist/
+            !dist/dbt-${{ inputs.version_number }}.tar.gz
+            ${{ needs.build-changelog-path.outputs.changelog_path }}
+          if-no-files-found: error
+          retention-days: ${{ env.artifact_retention_days }}
+
+      - name: Configure AWS credentials from Test account
+      # TODO: do i need to do this once per job?
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-east-1
+
+      - name: Upload Artifacts to S3 via CLI
+        run: |
+          aws s3 cp ${{ needs.build-changelog-path.outputs.changelog_path }} ${{ env.s3_bucket }}/dist --recursive  # since it's an entire directory
+
+      - name: Upload Changelog to S3 via CLI
+        run: |
+          aws s3 cp ${{ needs.build-changelog-path.outputs.changelog_path }} ${{ env.s3_bucket }}/${{ needs.build-changelog-path.outputs.changelog_path }}
+
+      - name: If we made it here we need to test what we built
+        id: set_success
+        run: echo '::set-output name=finished::true'
+      
+  test-build:
```

Comment:
I would think about this flow a little differently here but let me know what you think. Once the artifacts are in s3, I would consider them ready to ship and in 99.99% of all cases, those are what we ship. Testing the artifacts after we put them in S3 seems late to me especially if testing fails then we are in a weird state where we need to debug and clean-up s3. I would instead think the flow would be build -> test -> upload to s3 instead of build -> upload to s3 -> test. Then in the case where we come into this workflow and see that s3 already has the artifacts, we just skip the rest of the steps b/c we are good to go

Reply:
So much cleaner!  Love when someone with fresh eyes can rearrange code you stared at for too long.